crates/provider/src/heart.rs

```diff
@@ -415,37 +413,69 @@ impl<S> Heartbeat<S> {
         trace!(?to_watch.config);
+
         if let Some(timeout) = to_watch.config.timeout {
             self.reap_at.insert(Instant::now() + timeout, to_watch.config.tx_hash);
         }
+
+        // Start polling for the receipt.
+        let _ = hash_tx.send(to_watch.config.tx_hash).await;
         self.unconfirmed.insert(to_watch.config.tx_hash, to_watch);
     }
 
-    /// Handle a new block by checking if any of the transactions we're
-    /// watching are in it, and if so, notifying the watcher. Also updates
-    /// the latest block.
-    fn handle_new_block(&mut self, block: Block, latest: &watch::Sender<Option<Block>>) {
-        // Blocks without numbers are ignored, as they're not part of the chain.
-        let Some(block_height) = &block.header.number else { return };
+    /// Processes a receipt response. If the receipt is not yet available, sends new request.
+    /// Otherwise, either notifies the watcher or adds transaction to the waiting list.
+    async fn handle_receipt_response(
+        &mut self,
+        tx_hash: B256,
+        receipt: TransportResult<Option<N::ReceiptResponse>>,
+        hash_tx: &mpsc::Sender<B256>,
+    ) {
+        if !self.unconfirmed.contains_key(&tx_hash) {
+            // This likely means the transaction was reaped.
+            return;
+        }
+        if let Ok(Some(receipt)) = receipt {
+            let watcher = self.unconfirmed.remove(&tx_hash).unwrap();
 
-        // Check if we are watching for any of the transactions in this block.
-        let to_check =
-            block.transactions.hashes().filter_map(|tx_hash| self.unconfirmed.remove(tx_hash));
-        for watcher in to_check {
             // If `confirmations` is not more than 1 we can notify the watcher immediately.
             let confirmations = watcher.config.required_confirmations;
             if confirmations <= 1 {
-                watcher.notify();
-                continue;
+                return watcher.notify(receipt);
+            }
+
+            if let Some(block) = receipt.block_number() {
+                // Otherwise add it to the waiting list.
+                debug!(tx=%watcher.config.tx_hash, %block, confirmations, "adding to waiting list");
+                self.waiting_confs
+                    .entry(block + confirmations - 1)
+                    .or_default()
+                    .push((watcher, receipt));
+            } else {
+                // Edge case: receipt does not include block number, we can't count confirmations,
+                // so notify immediately.
+                watcher.notify(receipt);
             }
-            // Otherwise add it to the waiting list.
-            debug!(tx=%watcher.config.tx_hash, %block_height, confirmations, "adding to waiting list");
-            self.waiting_confs.entry(*block_height + confirmations - 1).or_default().push(watcher);
+        } else {
+            // Keep polling
+            let _ = hash_tx.send(tx_hash).await;
```

Comment:
this sends the hash back to the receipt fetcher task?

I find the naming here a bit confusing